server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java

```diff
@@ -170,41 +173,98 @@ public void onNewInfo(ClusterInfo info) {
                     reroute = true;
                     explanation = "high disk watermark exceeded on one or more nodes";
+                    usagesOverHighThreshold.add(usage);
+                    // will log about this node when the reroute completes
                 } else {
                     logger.debug("high disk watermark exceeded on {} but an automatic reroute has occurred " +
                             "in the last [{}], skipping reroute",
                         node, diskThresholdSettings.getRerouteInterval());
                 }
-                nodeHasPassedWatermark.add(node);
+
             } else if (usage.getFreeBytes() < diskThresholdSettings.getFreeBytesThresholdLow().getBytes() ||
                 usage.getFreeDiskAsPercentage() < diskThresholdSettings.getFreeDiskThresholdLow()) {
-                nodeHasPassedWatermark.add(node);
+
+                nodesOverHighThresholdAndRelocating.remove(node);
+
+                final boolean wasUnderLowThreshold = nodesOverLowThreshold.add(node);
+                final boolean wasOverHighThreshold = nodesOverHighThreshold.remove(node);
+                assert (wasUnderLowThreshold && wasOverHighThreshold) == false;
+
+                if (wasUnderLowThreshold) {
+                    logger.info("low disk watermark [{}] exceeded on {}, replicas will not be assigned to this node",
+                        diskThresholdSettings.describeLowThreshold(), usage);
+                } else if (wasOverHighThreshold) {
+                    logger.info("high disk watermark [{}] no longer exceeded on {}, but low disk watermark [{}] is still exceeded",
+                        diskThresholdSettings.describeHighThreshold(), usage, diskThresholdSettings.describeLowThreshold());
+                }
+
             } else {
-                if (nodeHasPassedWatermark.contains(node)) {
-                    // The node has previously been over the high or
-                    // low watermark, but is no longer, so we should
-                    // reroute so any unassigned shards can be allocated
-                    // if they are able to be
+
+                nodesOverHighThresholdAndRelocating.remove(node);
+
+                if (nodesOverLowThreshold.contains(node)) {
+                    // The node has previously been over the low watermark, but is no longer, so it may be possible to allocate more shards
+                    // if we reroute now.
                     if (lastRunTimeMillis.get() < currentTimeMillis - diskThresholdSettings.getRerouteInterval().millis()) {
                         reroute = true;
                         explanation = "one or more nodes has gone under the high or low watermark";
-                        nodeHasPassedWatermark.remove(node);
+                        nodesOverLowThreshold.remove(node);
+                        nodesOverHighThreshold.remove(node);
+
+                        logger.info("low disk watermark [{}] no longer exceeded on {}",
```

Comment:
I think that if we go from above high threshold to below low threshold, we log this message (or the debug message)?
Given that we log that high disk watermark is no longer exceeded when going below high threshold (but not low), I think we should also check here if we were previously above high threshold and if so, log that high disk watermark (or both low and high) is no longer exceeded?

Also, this message can be delayed up to the reroute interval (plus cluster info update interval), would it make sense to output it as soon as we know that disk usage has recovered instead?

Reply:
I added assertions about crossing multiple watermarks in ff041dc. I think it's unnecessary to try and describe exactly which watermarks have been crossed: we require them to be in order so if you are no longer exceeding the low watermark then you are also not exceeding any higher watermarks too.